Add ProtectKernelLogs to systemd.exec

keur opened the pull request.

boucman commented:
can we have some documentation ?

keur commented:
@boucman yeah, will add this tomorrow.

poettering commented on src/shared/seccomp-util.c:
```diff
@@ -523,6 +523,12 @@ const SyscallFilterSet syscall_filter_sets[_SYSCALL_FILTER_SET_MAX] = {
+                .help = "Kernel message ring buffer",
+                .value =
+                "syslog\0"
+        },
```
hmm, having a group for a single syscall only sounds unnecessary. Let's just block the syscall directly, no need to go via the group

poettering commented:
Looks good mostly, would prefer without the syscall group definition. Groups only make sense if we need to block multiple syscalls, but in this case it's just one, in which case we should just block it directly without going via the indirection of the syscall group.

Hmm, this should probably also block access to /dev/kmsg. Originally I thought this would mean hooking up things with DeviceAllow=, but given that that's a whitelist anyway, it makes little sense. Also, blocking /dev/kmsg is probably not enough anyway, we should also block /proc/kmsg. Hence, I think this should be hooked to the namespacing logic instead: i.e. extend NamespaceInfo with one more boolean and then overmount both /dev/kmsg and /proc/kmsg like we do for ProtectKernelTunables= to make it go away, and leave the DeviceAllow= infrastructure untouched.

keur commented:
Changes made cc @poettering 

Documentation added cc @boucman @anitazha

boucman commented:
isn't syslog about normal logging, not kernel logging ?

boucman commented:
Ok, there is a syslog kernel call and a syslog libc call.

This is about the kernel call, so it's fine... maybe worth clarifying in the documentation ? (not really your problem, but probably confusing for most people)

keur commented:
On Mon Nov 11, 2019 at 12:08 AM, Boucman wrote:
> Ok, there is a syslog kernel call and a syslog libc call.
>
>
> This is about the kernel call, so it's fine... maybe worth clarifying in
> the documentation ? (not really your problem, but probably confusing for
> most people)
>

I referred to it as the "syslog system call" in the documentation. I was
trying to be verbose to avoid this confusion, but it seems I have missed
the mark. Any suggestions?

boucman commented:
no, I had missed that but it's probably explicit enough... I was confused by the fact that we have a syscall which is named exactly like a libc call but does something completely different

poettering commented on man/systemd.exec.xml:
```diff
@@ -1321,6 +1321,19 @@ BindReadOnlyPaths=/var/lib/systemd</programlisting>
+      <varlistentry>
+        <term><varname>ProtectKernelLogs=</varname></term>
+
+        <listitem><para>Takes a boolean argument. If true, access to the kernel ring buffer will be denied. It is
```
kernel ring buffer → kernel log ring buffer

poettering commented on man/systemd.exec.xml:
```diff
@@ -1321,6 +1321,19 @@ BindReadOnlyPaths=/var/lib/systemd</programlisting>
+        <term><varname>ProtectKernelLogs=</varname></term>
+
+        <listitem><para>Takes a boolean argument. If true, access to the kernel ring buffer will be denied. It is
+        recommended to turn this on for most services that do not need to read kernel logs. Enabling this option
```
to read kernel logs → to read from or write to the kernel log ring buffer

poettering commented on man/systemd.exec.xml:
```diff
@@ -1321,6 +1321,19 @@ BindReadOnlyPaths=/var/lib/systemd</programlisting>
+        recommended to turn this on for most services that do not need to read kernel logs. Enabling this option
+        removes <constant>CAP_SYSLOG</constant> from the capability bounding set for this unit, and installs a system
+        call filter to block the <citerefentry project='man-pages'><refentrytitle>syslog</refentrytitle><manvolnum>2</manvolnum></citerefentry>
+        system call. The kernel exposes logs to userspace via <filename>/dev/kmsg</filename> and
```
"system call" → "system call (not to be confused by the libc API syslog(3) for userspace logging)"

poettering commented on man/systemd.exec.xml:
```diff
@@ -1321,6 +1321,19 @@ BindReadOnlyPaths=/var/lib/systemd</programlisting>
+        removes <constant>CAP_SYSLOG</constant> from the capability bounding set for this unit, and installs a system
+        call filter to block the <citerefentry project='man-pages'><refentrytitle>syslog</refentrytitle><manvolnum>2</manvolnum></citerefentry>
+        system call. The kernel exposes logs to userspace via <filename>/dev/kmsg</filename> and
+        <filename>/proc/kmsg</filename>. If enabled, these are made inaccessible to all the processes in the unit.</para>
```
exposes logs → exposes its log buffer

(background: the logs are eventually available through other means because journald for example write them to disk. hence, this option restricts access to the log buffer not the logs so much, since one is an interface the other just a concept of information, if you follow what i mean. — i do agree though that the whole option should still be named ProtectKernelLogs= [and not ProtectKernelLogBuffer=] since this distinction is good to make in prose but appears a bit too philosophical to expose in the option, which should be named brief I think)

poettering commented on src/shared/seccomp-util.c:
```diff
@@ -1281,6 +1281,37 @@ int seccomp_protect_sysctl(void) {
+
+                if (r < 0) {
+                        log_debug_errno(r, "Failed to add syslog() rule for architecture %s, skipping %m", seccomp_arch_to_string(arch));
+                }
```
please do a "continue" here... if there's nothing in the filter there's no point in activating it

poettering commented on src/test/test-seccomp.c:
```diff
@@ -322,6 +322,42 @@ static void test_protect_sysctl(void) {
+
+#ifdef __NR_syslog
+                assert_se(syscall(__NR_syslog, 0, 0, 0) < 0);
+                assert_se(errno == EPERM);
```
maybe also check that open("/dev/kmsg", O_RDONLY) fails with EPERM and didn't before. (some minor additional care should be taken though: these CI tests might run in container environments, where the device is generally not available, hence if the opening didn't work beforehand, then just skip this open() check)

keur replied:
I've made all the other changes, but not sure about this one. Should this check really be in the seccomp tests? `seccomp_protect_syslog()` is not going to  restrict access there.

poettering replied:
sure, test for that can be elsewhere, but it should be there, after all the path /dev/kmsg was misspelled...

keur commented:
Changes made cc @poettering 

Also had to make a small tweak to the test because inside the container `syslog` will `EPERM`. It was causing CI to fail.

poettering commented on src/core/namespace.c:
```diff
@@ -109,6 +109,12 @@ static const MountEntry protect_kernel_modules_table[] = {
+/* ProtectKernelLogs= option */
+static const MountEntry protect_kernel_logs_table[] = {
+        { "/proc/kmsg",          INACCESSIBLE, true },
+        { "/dev/kmesg",          INACCESSIBLE, true },
```
/dev/kmesg → /dev/kmsg

(a test would have caught this?)

poettering replied:
unfixed

poettering commented:
Looks excellent! Except for that one typo, which made me really really want a test for the namespacing part...

poettering commented:
Two more things I just noticed: this should have some hookup in "systemd-analyze security". And we should turn this on for all our long-running services. (Not systemd-udevd though and not systemd-journald, since theyneed to create/interface with /dev/kmsg)

keur commented:
Changes made. 

- Added a test (only for `/dev/kmsg`, because `/proc/kmsg` is still readable, just gets an empty file and I couldn't think of a good way to do a before/after for that one).

- Added `ProtectKernelLogs=yes` to the long-running units

- Added `ProtectKernelLogs` to `systemd-analyze`.

cc @poettering

poettering commented on src/core/namespace.c:
```diff
@@ -112,7 +112,7 @@ static const MountEntry protect_kernel_modules_table[] = {
 static const MountEntry protect_kernel_logs_table[] = {
         { "/proc/kmsg",          INACCESSIBLE, true },
-        { "/dev/kmesg",          INACCESSIBLE, true },
+        { "/dev/kmsg",           INACCESSIBLE, true },
```
we generally prefer merging "perfect" patch sets, i.e. patchsets that are perfectly bisectable. Hence please fold this fixup into the original patch that broke it. Or in other words, in the same PR there shouldn't be patches that that break and then others that fix it.

poettering commented on src/test/test-namespace.c:
```diff
@@ -2,6 +2,7 @@
 
 #include <sys/socket.h>
 #include <sys/stat.h>
+#include <fcntl.h>
```
doesn't matter much, but please order alphabetically

poettering commented on src/test/test-namespace.c:
```diff
@@ -102,7 +106,61 @@ static int test_netns(void) {
+        if (pid == 0) {
+                _cleanup_close_ int fd = -1;
+
+                fd = open("/dev/kmsg", O_RDONLY);
```
doesn#t matter much in this case, but we made it a habit to always specifiy O_CLOEXEC unless there's explicit reason not to, so that grepping over all open()s tells us quickly that we didn't forget it anwhere.

poettering commented on src/test/test-namespace.c:
```diff
@@ -102,7 +106,61 @@ static int test_netns(void) {
+                                    NULL);
+                assert_se(r == 0);
+
+                assert_se(open("/dev/kmsg", O_RDONLY) < 0);
```
dito

poettering commented on src/test/test-namespace.c:
```diff
@@ -133,5 +191,6 @@ int main(int argc, char *argv[]) {
 
-        return test_netns();
+        test_netns();
+        test_protect_kernel_logs();
```
please add "return EXIT_SUCCESS" to the end, so that main() returns something clean

poettering commented:
looks great, we are getting close!

keur commented:
Changes made. I changed `__NR_syslog` in `test-seccomp` to have the form of https://github.com/systemd/systemd/pull/14032

cc @poettering

poettering commented:
CI failed:

```
====== test-namespace =======
Assertion 'errno == EACCES' failed at src/test/test-namespace.c:158, function test_protect_kernel_logs(). Aborting.
ns-kernellogs terminated by signal ABRT.
Assertion 'wait_for_terminate_and_check("ns-kernellogs", pid, WAIT_LOG) == EXIT_SUCCESS' failed at src/test/test-namespace.c:163, function test_protect_kernel_logs(). Aborting.
FAIL: test-namespace (code: 134)
```

poettering commented:
Looks otherwise great!

So the INACCESSIBLE namespacing stuff just overmounts the device node with an empty file that is marked with access mode 000... because of that it might be better to drop privs before checking whether we can read it in the test. i.e. `assert_se(setresuid(UID_NOBODY, UID_NOBODY, UID_NOBODY) >= 0` after applying the namespace stuff, but before open()ing the device. i.e. that way the effect of the access mode 000 takes full effect. (in contrast to staying root, since for root the access mode of files generally doesn't matter).

keur commented:
@poettering Good call. I pushed that. Let's see if that fixes CI. Will check back tomorrow morning.

poettering commented:
neat! all passed! thank you very much for this work!

Geass-LL commented:
Until this commit, if we run `while true; do systemctl daemon-reload; done`, the Resident Set Size(RSS) systemd used will increase 
gradually. (In my case, 8K/min). So I think there may be some minor memory leak when daemon-reload. However, after merging patches in this issue, the RSS stops increasing.

Geass-LL commented:
```shell
while true; do
    systemctl daemon-reload;
done
```
```shell
while true;do
    nowtime=`date`
    res=`ps aux | grep "systemd --switched-root --system --deserialize" | grep -v grep`
    echo "${nowtime}: ${res}" >> /root/log
    sleep 30
done
```
My shell scripts, may be useful.
